### PR TITLE
Hw2ex

### PR DESCRIPTION
**Pull request by NAlexH2**
I've taken some of the comments you all left and added it into this round of code. You may see me being inconsistent between the exercises but it's because I am trying different things to see what works best or looks good.

* Completed all of CSS Grid Garden (ex1)
* Completed ex2 and learned plenty of new things (especially with highlighting and regex!)
* Followed up the knowledge I got in ex2 into ex3 and created the highlight for that fixed field of information. Learned in both how to use the `<mark>` element.
    * You'll notice I modify the style in both, but that's because the default color for `<mark>` is *awful*
* Learned about intervals with ex4 and how to use them. Wound up giving up on a particular solution and going with what you see now. I was trying to pass around an int, but it didn't work, so I made a tiny little array to push pop from instead.
* ex5 actually got me more familiar and comfortable with Bootstrap than any of the others did, simply because of having to support the imports and manipulate the modal directly.

- [x] Ex1 - CSS Grid Garden
- [x] Ex2 - Search for names and highlight
- [x] Ex3 - Search the message and highlight
- [x] Ex4 - Random background
- [x] Ex5 - Form

### CODE REVIEW

File: hw2/02-search.html [hunk 1]
```diff
@@ -1,23 +1,48 @@
...
+        <link
+            rel="stylesheet"
+            href="https://cdn.jsdelivr.net/npm/bootstrap@5.0.2/dist/css/bootstrap.min.css"
+        />
+        <script src="02-data.js"></script>
+        <style>
+            mark {
+                background-color: yellow; /* Change this to the desired highlight color */
+            }
+        </style>
```
Comment (glatimer) [inline hw2/02-search.html:16]: Where did you find this mark element?  I haven't been able to find it.  What does it do?

File: hw2/03-count.html [hunk 2]
```diff
@@ -1,34 +1,38 @@
...
+            }
+        </style>
+    </head>
+    <body
+        class="flex justify-content-center align-content-center row"
+        style="background-color: rgb(0, 66, 92)"
+    >
+        <!-- The blue does not exist within the bootstrap classnames -->
+        <main class="col-6 bg-light container mt-5 rounded">
+            <div class="">
```
Comment (glatimer) [inline hw2/03-count.html:23]: I think you meant to add a class name here?  Either way, this div could use some padding to add a little white space between the text and the border.

File: hw2/03-count.js [hunk 3]
```diff
@@ -1,3 +1,62 @@
...
+epoch of belief, it was the epoch of incredulity, it was the
+season of Light, it was the season of Darkness, it was the
+spring of hope, it was the winter of despair, we had
+everything before us, we had nothing before us, we were all
+going direct to Heaven, we were all going direct the other
+way—in short, the period was so far like the present period,
+that some of its noisiest authorities insisted on its being
+received, for good or for evil, in the superlative degree of
+comparison only.
+`;
```
Comment (glatimer) [inline hw2/03-count.js:16]: I can imagine your const variable here would be more efficient as well.  Just make sure the HTML page content never changes!

File: hw2/03-count.js [hunk 4]
```diff
@@ -1,3 +1,62 @@
...
+input.addEventListener("keydown", handleKeyDown);
+
+// Access the most modified element in the doc with ease, again!
+const getContElement = function getContElement() {
+    return document.getElementById("toSearch");
+};
+
+// Very easy regex function to highlight words by modifying the inner html.
+// I'd rather *not* do this, but I think it's okay. Doing it any other way would
+// not be fun. If the regex fails just reset the search.
```
Comment (glatimer) [inline hw2/03-count.js:36]: haha I did the same thing!  Curious to know why innerHTML is such a no no. Will look into it.

File: hw2/04-background.js [hunk 5]
```diff
@@ -1 +1,55 @@
...
+});
+
+// used to stop the interval
+const intervalID = [];
+
+// add a click listener to the button
+document.getElementById("actionButton").addEventListener("click", buttonListen);
+
+// Neatly generate the string for the random color to modify the style
+const randColor = function getRandomColor() {
```
Comment (glatimer) [inline hw2/04-background.js:14]: Had no idea this getRandomColor() function existed. I need to look at the libraries more.  I'm so used to building everything from scratch.

File: hw2/05-form.html [hunk 6]
```diff
@@ -1,20 +1,141 @@
...
+        />
+        <script src="https://code.jquery.com/jquery-3.6.0.min.js"></script>
+        <script src="https://cdn.jsdelivr.net/npm/bootstrap@5.0.2/dist/js/bootstrap.bundle.min.js"></script>
+        <script defer src="05-form.js"></script>
+    </head>
+    <body class="bg-dark">
+        <main class="col-6 bg-light container mt-5 rounded shadow ps-3">
+            <h1 class="text-center fw-bold pt-3 pb-2">Exercise 05 - Form</h1>
+            <form id="dataForm" class="py-3 w-100">
+                <label class="w-100 fw-bold pb-1 rounded-3" for="fullName"
```
Comment (glatimer) [inline hw2/05-form.html:20]: Consider using class="form-label" to ensure accessibility through bootstraps built in styling.

File: hw2/05-form.html [hunk 7]
```diff
@@ -1,20 +1,141 @@
...
+                        <option value="" hidden selected>
+                            Choose an Option
+                        </option>
+                        <option value="Registered">Registered</option>
+                        <option value="Withdrawn">Withdrawn</option>
+                    </select>
+                    <div class="fw-bold pb-1 pt-3 rounded-3">
+                        Have you taken any of the following courses?
+                    </div>
+                    <div>
```
Comment (glatimer) [inline hw2/05-form.html:62]: Consider adding a fieldset around your checkboxes to add more semantical structure for accessibility.

**Review by glatimer — APPROVED** [review]
I learned I could open the project in codepen to inspect the elements and it was helpful for giving a more informed review. Good job applying comments from last week!  This looks good.

File: hw2/02-search.html (hunk 1 above)
Comment (NAlexH2) [reply hw2/02-search.html:16]: it highlights text that is wrapped in the element.

File: hw2/04-background.js (hunk 5 above)
Comment (NAlexH2) [reply hw2/04-background.js:14]: It's not, I just wrote that function and gave it a const shortened name to give it structured hoisting.

File: hw2/02-search.html [hunk 8]
```diff
@@ -1,23 +1,48 @@
...
+        <meta name="viewport" content="width=device-width, initial-scale=1.0" />
+        <title>02 - Search</title>
+        <link
+            rel="stylesheet"
+            href="https://cdn.jsdelivr.net/npm/bootstrap@5.0.2/dist/css/bootstrap.min.css"
+        />
+        <script src="02-data.js"></script>
+        <style>
+            mark {
+                background-color: yellow; /* Change this to the desired highlight color */
```
Comment (PhatCash) [inline hw2/02-search.html:14]: This is great. The default color is very light.

File: hw2/02-search.js [hunk 9]
```diff
@@ -1,3 +1,85 @@
...
+// Create new cards for every character provided. Doing this by creating a new
+// 'floating' element then appending them in the correct order before
+// conducting work. Using regex to do the replace for the mark html5 element
+// See my 02-search.html for the style mod for the mark element to make it
+// stand out more.
+const characterCards = function buildCharacterCards(relevantChars, searchVal) {
+    const resultElem = getResElem();
+    const regex = new RegExp(`${searchVal}`, "gi");
+    for (const character of relevantChars) {
+        const cleanCards = document.createElement("div");
```
Comment (PhatCash) [inline hw2/02-search.js:45]: This approach is safer than using ```innerHTML```, so thank you for pointing that out.

File: hw2/03-count.html [hunk 10]
```diff
@@ -1,34 +1,38 @@
...
+        />
+        <style>
+            mark {
+                background-color: yellow; /* Change this to the desired highlight color */
+            }
+        </style>
+    </head>
+    <body
+        class="flex justify-content-center align-content-center row"
+        style="background-color: rgb(0, 66, 92)"
```
Comment (PhatCash) [inline hw2/03-count.html:19]: That's correct. I couldn't find the blue from the example unless adding a custom color as you did here.

File: hw2/04-background.html [hunk 11]
```diff
@@ -1,29 +1,45 @@
...
+            <div>
+                <label for="userInput"
+                    >Background change interval (seconds):
+                </label>
+            </div>
+            <div class="">
+                <input
+                    id="userInput"
+                    type="number"
+                    step="1"
```
Comment (PhatCash) [inline hw2/04-background.html:30]: Good use of the min and step attributes for the input element. It guards the user from entering negative numbers and allows increments of one.

File: hw2/05-form.js [hunk 12]
```diff
@@ -1 +1,123 @@
...
+// Way easier to hide it than showing it...
+function hideModal() {
+    document.getElementById("modalBody").innerHTML = "";
+    myModal.hide();
+}
+
+// Write a few ternaries so they are easier to read on their own.
+// Each either returns the value in the input (the class name) or it returns
+// an empty string. We check/use this later when building the modal.
+const getPL = function getProgramingLanguage() {
```
Comment (PhatCash) [inline hw2/05-form.js:85]: This highlights the power of using the ternary operator. As you mentioned, the Airbnb documentation states in rule **15.6** to not be nested and be single-line expressions. I know that prettier automatically formats them this way and I find it easier to read them like this over single-line expressions.

**Review by PhatCash — APPROVED** [review]
Reviewed. Nice job!

File: hw2/02-search.js (hunk 9 above)
Comment (NAlexH2) [reply hw2/02-search.js:45]: It is but it's not pretty haha. It took some serious wrangling to figure it out and get it just right.

File: hw2/03-count.html (hunk 10 above)
Comment (NAlexH2) [reply hw2/03-count.html:19]: Wound up using windows power toys and using the color picker tool it has to get its values.

File: hw2/05-form.js (hunk 12 above)
Comment (NAlexH2) [reply hw2/05-form.js:85]: Agreed! This certainly is easier to read, but you can't always get away with a single line too. I suspect what someone would say who wrote that style guide would be like "if you can't fit it into a single line, just turn it into an if/else" but then you're just sitting there with even more lines of code that do the exact same thing. Prettier taking over and formatting it nicely is really great here.

File: hw2/02-search.js [hunk 13]
```diff
@@ -1,3 +1,85 @@
 // Add your code here
 
-userInput.addEventListener('click', handleClick);
+document.getElementById("searchButton").addEventListener("click", handleClick);
```
Comment (baylie-r-g) [inline hw2/02-search.js:3]: I see you also noticed the error in the original event listener. Caused some errors before I figured that one out

File: hw2/02-search.js [hunk 14]
```diff
@@ -1,3 +1,85 @@
...
+const getResElem = function getResultElement() {
+    return document.getElementById("searchResult");
+};
+
+// simple quick empty check
+const checkEmpty = function checkInputEmpty(value) {
+    return value.length === 0;
+};
+
+// If the search was empty, let em know.
```
Comment (baylie-r-g) [inline hw2/02-search.js:16]: I didn't think to check for this case. Well written

File: hw2/03-count.html [hunk 15]
```diff
@@ -1,34 +1,38 @@
...
+                <div>
+                    <label for="userInput"
+                        >Highlight all occurrences of a word:</label
+                    >
+                    <input type="text" id="userInput" />
+                </div>
+                <hr />
+                <div id="toSearch"></div>
+            </div>
+            <script defer src="03-count.js"></script>
```
Comment (baylie-r-g) [inline hw2/03-count.html:35]: Interesting choice to move the text to the js file. I think that is a smart move regarding ease of use.

File: hw2/03-count.js (hunk 4 above)
Comment (baylie-r-g) [reply hw2/03-count.js:36]: I agree that it was likely the best move. This is what I started doing before running out of time on the assignment

File: hw2/04-background.html (hunk 11 above)
Comment (baylie-r-g) [reply hw2/04-background.html:30]: @PhatCash I definitely didn't think of that, great job on hitting those edge cases.

**Review by baylie-r-g — APPROVED** [review]
Looks great Alex! Written quite clearly and concisely. I feel that I can likely learn a lot by studying your code

**Pull request merged**